src/elm/overloaded.js

```diff
@@ -60,19 +60,25 @@ class Union extends Expression {
   exec(ctx) {
     const [a, b] = this.execArgs(ctx);
     if (a == null && b == null) {
-      return null;
+      return this.listTypeArgs() ? [] : null;
     }
     if (a == null || b == null) {
       const notNull = a || b;
       if (typeIsArray(notNull)) {
         return notNull;
       } else {
-        return [];
+        return null;
       }
     }
     const lib = typeIsArray(a) ? LIST : IVL;
     return lib.doUnion(a, b);
   }
+
+  listTypeArgs() {
+    return this.args.every(arg => {
+      return arg.asTypeSpecifier != null && arg.asTypeSpecifier.type === 'ListTypeSpecifier';
+    });
```

Comment:
Since the allowed arguments to `union` are either `List, List` or `Interval, Interval`, I don't think it's necessary to check `every` arg.  I think checking for `some` arg would be sufficient and would actually be better, since cases where _both_ args are cast are probably contrived.